keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,872 @@
+    other pods but daemonset pods. This is tricky to do and error prone,
+    especially for smaller nodes. This will not be possible if node size is not
+    known in advance.
+-   Use pod anti affinity. Users will need to define the same set of labels on
+    all pods that needs to be repelled, whether or not they actually request 1:1
+    pod to node assignment, it is also expensive to calculate especially for
+    cluster autoscaler.
+
+A simple and native way to force 1:1 pod-to-node assignment is therefore needed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod
+
+### Non-Goals
+
+-   Find the best fit node for a given pod
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new pod-level API that allows users to specify how many "worker" pods are
+allowed on a node. The API is implemented by the scheduler as a filter plugin
+and kubelet as an admission predicate.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+#### Story 2
+
+As a user, I want to deploy a Spark or MPI job such that only a single worker is
+running on each pod excluding daemonsets.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Users might be able to circumvent 1:1 pod to node assignments by creating a
+daemonset with node affinity to a specific node.
+
+#### Possible Mitigations
+
+-   Prevent users from creating daemonsets using a
+    [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+    -   Object count quota (`count/daemonsets`) can be used for this purpose.
+-   Restrict daemonsets creation to admins/specific users via kubernetes
+    [RBAC](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
```

Comment:
it is fairly the same mitigation: prevent users from creating daemonsets, so merge them into one and mention that an admin could either use RBAC or ResourceQuota to do that.